images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go

```diff
@@ -0,0 +1,140 @@
+limitations under the License.
+*/
+
+package vmip
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmip/internal/util"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vmipcondition"
+)
+
+func NewVMIPValidator(log logr.Logger, client client.Client) *VMIPValidator {
+	return &VMIPValidator{log: log.WithName(controllerName).WithValues("webhook", "validation"), client: client}
+}
+
+type VMIPValidator struct {
+	log    logr.Logger
+	client client.Client
+}
+
+func (v *VMIPValidator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	vmip, ok := obj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualMachineIPAddress but got a %T", obj)
+	}
+
+	v.log.Info("Validate VirtualMachineIP creating", "name", vmip.Name, "type", vmip.Spec.Type, "address", vmip.Spec.StaticIP)
+
+	err := v.validateSpecFields(vmip.Spec)
+	if err != nil {
+		return nil, fmt.Errorf("error validating VirtualMachineIP creation: %w", err)
+	}
+
+	if vmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic {
+		ip := vmip.Spec.StaticIP
+
+		if ip != "" {
+			allocatedIPs, err := util.GetAllocatedIPs(ctx, v.client)
+			if err != nil {
+				return nil, fmt.Errorf("error getting allocated ips: %w", err)
+			}
+
+			allocatedLease, ok := allocatedIPs[ip]
+			if ok && allocatedLease.Spec.VirtualMachineIPAddressRef != nil && (allocatedLease.Spec.VirtualMachineIPAddressRef.Namespace != vmip.Namespace || allocatedLease.Spec.VirtualMachineIPAddressRef.Name != vmip.Name) {
+				return nil, fmt.Errorf("VirtualMachineIP cannot be created: the address %s has already been allocated for another VMIP", ip)
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func (v *VMIPValidator) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	oldVmip, ok := oldObj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected a old VirtualMachineIP but got a %T", oldObj)
+	}
+
+	newVmip, ok := newObj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualMachineIP but got a %T", newObj)
+	}
+
+	v.log.Info("Validate VirtualMachineIP updating", "name", newVmip.Name,
+		"old.type", oldVmip.Spec.Type, "new.type", newVmip.Spec.Type,
+		"old.address", oldVmip.Spec.StaticIP, "new.address", newVmip.Spec.StaticIP,
+	)
+
+	boundCondition, exist := service.GetCondition(vmipcondition.BoundType, oldVmip.Status.Conditions)
+	if exist && boundCondition.Status == metav1.ConditionTrue {
```

Comment:
This if tries to validate Bound state as Unbound state =) Consider simplifying nested ifs for Bound state and add another set of checks for Unbound state.

e.g. no need to check auto->static if types are non-equal in Bound state:
```
if oldVmip.Spec.Type != newVmip.Spec.Type {
return nil, fmt.Errorf("can't change type %s -> %s in the 'Bound' state", oldVmip.Spec.Type, newVmip.Spec.Type)
}
```